elastic/kibana
# [Enterprise Search] Add toasts support to FlashMessages component

## Summary

I **strongly** recommend following along [by-commit](https://github.com/elastic/kibana/pull/95981/commits), and hiding whitespace changes!

This PR adds the ability to flash success & error messages via EuiToasts instead of EuiCallOuts. Examples:

<img width="373" alt="" src="https://user-images.githubusercontent.com/549407/113201207-1988c500-921e-11eb-8971-2bda7b0e4f6b.png">

(Not in this PR, I whipped it up locally for a screenshot)
<img width="381" alt="" src="https://user-images.githubusercontent.com/549407/113202383-894b7f80-921f-11eb-8ed1-006a472c54da.png">

Example usage:

```js
import { flashSuccessToast } from '../../../shared/flash_messages';

// Simple string message
flashSuccessToast('You did a thing!');

// Advanced options
flashSuccessToast('You did a thing!', {
  text: 'Optional descriptive text (per EUI guidelines, do not put too much information in here)',
  toastLifeTimeMs: 5000,
  id: 'customId',
});
```

### Toasts guidelines

**PLEASE READ BEFORE USING**:

- Generally, we should prefer to use toasts primarily for short success messages with no actions.
- Error messages should _generally_ still prefer to use EuiCallOuts so that issues remain constant & visible to users. 
  - An exception to this is in this PR: polling error messages that recur every 5 seconds is a perfect use case for toasts, because we do not have to manually clear them if the polling is able to resume connectivity.
- Info/warning messages (fairly rare in our codebase) should still also prefer callouts, although we can evaluate this on a case-by-case basis.
- Messages with actions should also generally still prefer callouts, although this can also be evaluated on a case-by-case basis.
- The current default timeout for toasts is 5 seconds. This can be overridden per-toast as needed, although I strongly caution to consider your use-case when changing the timeout & consulting with our UX designers or the EUI team if creating either a very long or very short toast.

EUI docs: https://elastic.github.io/eui/#/display/toast/guidelines

### Also in this PR

- Cleaned up FlashMessagesLogic tests slightly
- a11y enhancement: Update our callout messages to be announced by screen readers

## QA

- Start Enterprise Search and Kibana
- Create a sample engine
- Go to http://localhost:5601/xyz/app/enterprise_search/app_search/engines/national-parks-demo/api_logs, and open your dev tools to track XHR calls
- Stop your Enterprise Search instance
- [ ] Confirm that an error toast appears for each failed XHR poll

### Checklist
